File: LeetCode/C/19-two-pointers-删除链表的倒数第n个节点.cpp
/*
 * @lc app=leetcode.cn id=19 lang=cpp
 *
 * [19] 删除链表的倒数第N个节点
 */
#include "main.h"
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
  public:
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        ListNode *p = head, *pre = nullptr, *cur = head;
        while (n--)
            p = p->next;
        while (p) {
            pre = cur;
            cur = cur->next;
            p = p->next;
        }
        if (pre != nullptr) {
            pre->next = cur->next;
            delete cur;
        } else {
            pre = head;
            head = head->next;
            delete pre;
        }
        return head;
    }
};
// @lc code=end

File: LeetCode/Go/19-two-pointers-删除链表的倒数第n个节点.go
/*
 * @lc app=leetcode.cn id=19 lang=golang
 *
 * [19] 删除链表的倒数第N个节点
 */
package leetcode

// @lc code=start
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	var p, pre, cur *ListNode = head, nil, head
	for ; n > 0; n-- {
		p = p.Next
	}
	for p != nil {
		pre = cur
		cur = cur.Next
		p = p.Next
	}
	if pre != nil {
		pre.Next = cur.Next
	} else {
		head = head.Next
	}
	return head
}

// @lc code=end
